sdks/python/apache_beam/io/parquetio.py

```diff
@@ -448,6 +451,14 @@ def __init__(
         is '-SSSSS-of-NNNNN' if None is passed as the shard_name_template.
       mime_type: The MIME type to use for the produced files, if the filesystem
         supports specifying MIME types.
+      max_records_per_shard: Maximum number of records to write to any
+        individual shard.
+      max_bytes_per_shard: Target maximum number of bytes to write to any
```

Comment:
Is there a way to avoid this excess?

Reply:
it's possible IMO, but this is the FileBasedSink responsibility to take care of that.

(basically it should be able to slice the buffer that is about to be written and split it into the file being closed, and a new one)

Reply:
given that this is pass through to existing structure, we can not make the change

Reply:
It's impossible to always avoid, as a single record may exceed the bytes-per-shard limit. Even if this is not the case, some file formats have a footer/trailer (e.g. checksums, listings, indices, delimiters...) that would put one over the limit even if the written records were under, so it's really file-format-dependent on how this can be achieved. 

The intent here is that shard not be "too big" which is generally flexible.